src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceActivationMaintainer.cs

```diff
@@ -0,0 +1,387 @@
+                 var act = kp.Value.Instances.FirstOrDefault();
+                 if (act.Key == null) return false;
+                 if (act.Value.RegistrationStatus == MultiClusterStatus.Doubtful) return true;
+                 return false;
+             }).Select(kp => kp.Key);
+
+            lock (lockable)
+            {
+                doubtfulGrains.AddRange(newdoubtful);
+            }
+        }
+
+        public void TrackDoubtfulGrains(GrainDirectoryPartition newstuff)
+        {
+            var newdoubtful = newstuff.GetItems().Where(kp =>
+            {
+                if (!kp.Value.SingleInstance) return false;
+                var act = kp.Value.Instances.FirstOrDefault();
+                if (act.Key == null) return false;
+                if (act.Value.RegistrationStatus == MultiClusterStatus.Doubtful) return true;
+                return false;
+            }).Select(kp => kp.Key);
+
+            lock (lockable)
+            {
+                doubtfulGrains.AddRange(newdoubtful);
+            }
+        }
+
+        // the following method runs for the whole lifetime of the silo, doing the periodic maintenance
+        protected override async void Run()
+        {
+            var globalConfig = Silo.CurrentSilo.OrleansConfig.Globals;
+            if (!globalConfig.HasMultiClusterNetwork)
+                return;
+
+            var myClusterId = globalConfig.ClusterId;
+
+            while (router.Running)
+            {
+                try
+                {
+                    await Task.Delay(period);
+                    if (!router.Running) break;
+
+                    logger.Verbose("GSIP:M running periodic check (having waited {0})", period);
+
+                    // examine the multicluster configuration
+                    var config = Silo.CurrentSilo.LocalMultiClusterOracle.GetMultiClusterConfiguration();
+
+                    if (config == null || !config.Clusters.Contains(myClusterId))
+                    {
+                        // we are not joined to the cluster yet/anymore. 
+                        // go through all owned entries and make them doubtful
+                        // this will not happen under normal operation
+                        // (because nodes are supposed to shut down before being removed from the multi cluster)
+                        // but if it happens anyway, this is the correct thing to do
+
+                        var allEntries = router.DirectoryPartition.GetItems();
+                        var ownedEntries = allEntries.Where(kp =>
+                        {
+                            if (!kp.Value.SingleInstance) return false;
+                            var act = kp.Value.Instances.FirstOrDefault();
+                            if (act.Key == null) return false;
+                            if (act.Value.RegistrationStatus == MultiClusterStatus.Owned) return true;
+                            return false;
```

Comment:
Yes, agreed. I am refactoring the filter on the Dictionary... which also eliminates the LINQ query.